File: binarytree/117-populating-next-right-pointers-ii/ConnectNextPointersII.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null){
            return null;
        }
        Node dummy = new Node(0);
        Node prev = dummy;
        Node current = root;
        while(current!=null){
            if(current.left!=null){
                prev.next = current.left;
                prev = prev.next;
            }
            if(current.right!=null){
                prev.next = current.right;
                prev  = prev.next;
            }
            current = current.next;
            if(current==null){
                current = dummy.next;
                dummy.next = null;
                prev = dummy;
            }
        }
        return root;
    }
}

File: binarytree/117-populating-next-right-pointers-ii/ConnectNextPointersII.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None
        dummy = Node(0)
        prev = dummy
        current = root

        while current:
            if current.left:
                prev.next = current.left
                prev = prev.next
            if current.right:
                prev.next = current.right
                prev = prev.next
            current = current.next
            if not current:
                current = dummy.next
                dummy.next = None
                prev = dummy
        return root
